### ISSUE TITLE
DB connection via URI does not work when password contains &]~^ symbols

### ISSUE THREAD

**Issue by RicardoRamirezR** [tpope/vim-dadbod#171]
Good day,

Given this password ```#X!12j999%&]vK~^SQX``` 

The  DB URI  that looks like: ```mysql://user:%23X!12j999%&]vK~^SQX@127.0.0.1/example```

I encoded # as %23

The connector seems to take the password up to the first & and DB gives:

```
[DBUI]
Error connecting to db example: DB exec error: 
mysql: [ERROR] unknown variable '&]vK~^SQX@127.0.0.1:5000/example=1'
```

I've tried to encode:

- % -> %25
- & -> %26
- ] -> %5D
- ~ -> %7E
- ^ -> %5E

With no luck

**Comment by tpope** [comment]
Try  with `:DB mysql://user:%23X!12j999%&]vK~^SQX@127.0.0.1/example SELECT 1`. Does that work? If so, it's a Dadbod UI issue, not a Dadbod issue.

**Comment by RicardoRamirezR** [comment]
> Try with `:DB mysql://user:%23X!12j999%&]vK~^SQX@127.0.0.1/example SELECT 1`. Does that work? If so, it's a Dadbod UI issue, not a Dadbod issue.

Thank you @tpope, same error message:

![db](https://github.com/tpope/vim-dadbod/assets/6526545/9b9c9ba7-4451-46b7-badb-d00981484b73)

**Comment by tpope** [comment]
```vim
:echo db#adapter#mysql#interactive('mysql://user:%23X!12j999%&]vK~^SQX@127.0.0.1/example')
['mysql', '-h', '127.0.0.1', '-u', 'user', '-p#X!12j999%&]vK~^SQX', 'example']
```

I don't have a MySQL install handy. What happens when you run the corresponding shell command?

```
mysql -h 127.0.0.1 -u user "-p#X!12j999%&]vK~^SQX" example
```

**Comment by RicardoRamirezR** [comment]
I got:

<img width="977" alt="Screenshot 2024-05-18 at 3 09 16 PM" src="https://github.com/tpope/vim-dadbod/assets/6526545/ae65015e-84cd-4d5f-bbe0-c92c4acd7fc1">

**Comment by RicardoRamirezR** [comment]
BUT when I encode %& with %25%26 as: ```mysql://user:%23X!12j999%25%26]vK~^SQX@127.0.0.1/example``` it works. I thought I had done all the substitutions before, I just rechecked and made the correction. Thank you for your time.

**Comment by tpope** [comment]
I don't fully understand what went wrong here, but you should indeed encode those, so I guess we have our solution.

**Comment by przepompownia** [comment]
`db#url#parse(url)` splits `url` with `&` which leads to unexpected results. 

I started from investigate what happens wrong, but then found `:h dadbod-urls` where I should start. Users have to escape URL-s.

A hint how to encode URLs could be helpful to Vim users (epsecially when sourced from .env files). For my Lua Nvim config I have piped the original URL through `:gsub('[^%w@.:/ ]', function(char) return string.format('%%%02x', string.byte(char)) end)` but not expect from it to be universal (and maybe there is a simpler alternative).
